### ISSUE TITLE
SimpleJavaMail should be more clear about exceptions thrown where

### ISSUE THREAD

**Issue by huffstler** [bbottema/simple-java-mail#208]
Going through [subclass SendMailClosure](https://github.com/bbottema/simple-java-mail/blob/develop/modules/simple-java-mail/src/main/java/org/simplejavamail/mailer/internal/mailsender/MailSenderImpl.java#L214) and its associated `run()` method, I see there are scenarios where an exception will bubble up, though that feature isn't immediately noticeable.

I was doing the following:

```java
void send(String message) {
        var postcard = EmailBuilder
            .startingBlank()
            .withPlainText(message)
            .buildEmail();
        if (fedex.validate(postcard)) {
            fedex.sendMail(postcard);
            logger.info("Mail has sent successfully.");
        } else {
            logger.error("Email was formatted incorrectly.");
        }
    }
```

I was getting problems trying to send email through the program, and saw errors in the logs, but no reasons why.  I had to go digging into the source code to find out what [was actually happening](https://github.com/bbottema/simple-java-mail/blob/develop/modules/simple-java-mail/src/main/java/org/simplejavamail/mailer/internal/mailsender/MailSenderImpl.java#L283). Going through the documentation I couldn't find anything referencing the `send` method throwing an `Exception` either.

Since I now know that an `Exception` can bubble up, I've handled it correctly in the code: 

```java
void send(String message) {
    var postcard = EmailBuilder
        .startingBlank()
        .withPlainText(message)
        .buildEmail();
    if (fedex.validate(postcard)) {
        try {
            fedex.sendMail(postcard);
        } catch (Exception ex) {
            logger.error("Caught a thrown exception from inside SimpleJavaMail.", ex.getCause());
        }
        logger.info("Mail has sent successfully.");
    } else {
        logger.error("Email was formatted incorrectly.");
    }
}
```

Which shows the real problem 

```
15:30:42.932 ERROR com.foo.bar.Emailer - Caught a thrown exception from inside SimpleJavaMail.
com.sun.mail.smtp.SMTPSendFailedException: 530 5.7.1 Client was not authenticated
... more blood below
```

In your opinion is this something that should be addressed, or is it working as intended? In my view, it'd be easier for the user of the library to expect catching `Exceptions` that are passed up. Is there an easy way to indicate this?

**Comment by bbottema** [comment]
Hi and thanks for the report!

> I was getting problems trying to send email through the program, and saw errors in the logs, but no reasons why.

That doesn't sound right; at least you should be getting an exception with a stacktrace in your logging, unless the exception was caught upstream and swallowed (or put out to a log file you didn't expect).

Btw, the documentation _does_ mention it on the website under Logging & Debugging:

![image](https://user-images.githubusercontent.com/2115718/57176718-95740a00-6e5b-11e9-9952-eaadbbf0c033.png)

However, I agree that the code and javadoc should indicate it better through some Javadoc and a `throws` clause, even though it is a runtime exception, not a checked exception. I'll have it in the code for release 6.0.0.

Sidenote: you are now catching `Exception`, but you should probably catch `MailException` instead.

**Comment by huffstler** [comment]
> That doesn't sound right; at least you should be getting an exception with a stacktrace in your logging, unless the exception was caught upstream and swallowed (or put out to a log file you didn't expect).

I wouldn't be surprised if this was the case, I'll take a look around and see if the logs got dropped anywhere.

I totally missed that part of the docs on the website, my fault for not doing due diligence!

> Sidenote: you are now catching Exception, but you should probably catch MailException instead.

Consider it changed 👍 

Thanks for the quick response and suggestions!  I really appreciate it :)

**Comment by huffstler** [comment]
It turns out a `MailException` wasn't getting thrown, it's throwing an `Exception`. [This is the block that I'm hitting](https://github.com/bbottema/simple-java-mail/blob/develop/modules/simple-java-mail/src/main/java/org/simplejavamail/mailer/internal/mailsender/MailSenderImpl.java#L283-L286), the catch statement is handling the `Exception` but only logging details and re-throwing the same `Exception`.

``` java
try {
  fedex.sendMail(postcard);
  logger.info("Mail has sent successfully.");
} catch (MailException ex) {
  logger.error("Error sending email...", ex.getCause())
}
```
Stacktrace from the above code (sensitive info removed): 

```
11:19:00.125 DEBUG org.simplejavamail.mailer.internal.mailsender.MailSender - starting mail with session (host: mail.wsfsbank.com, port: 587, username: null, authenticate: null, transport: SMTP_TLS)
11:19:05.348 ERROR org.simplejavamail.mailer.internal.mailsender.MailSender - Failed to send email:
Email{
	id=<2092885124.3.1557242340141@foo.com>
	fromRecipient=Recipient{name='foo', address='bar@baz.com', type=null},
	replyToRecipient=null,
	bounceToRecipient=null,
	text='Couldn't call stored procedure on file.csv',
	textHTML='null',
	subject='[TEST POST PLS IGNORE]',
	recipients=[Recipient{name='null', address='me@foo.com', type=To}]
}
```

```java
try {
  fedex.sendMail(postcard);
  logger.info("Mail has sent successfully.");
} catch (Exception ex) {
  logger.error("Error sending email...", ex.getCause())
}
```
Stacktrace from the above code (sensitive data removed): 
```
09:54:58.727 DEBUG org.simplejavamail.mailer.internal.mailsender.MailSender - starting mail with session (host: smtp.foo.com, port: 587, username: null, authenticate: null, transport: SMTP_TLS)
09:55:13.504 ERROR org.simplejavamail.mailer.internal.mailsender.MailSender - Failed to send email:
Email{
	id=<2126723403.3.1557150903269@foo.net>
	fromRecipient=Recipient{name='me', address='me@me.com', type=null},
	replyToRecipient=null,
	bounceToRecipient=null,
	text='This is a test message',
	textHTML='null',
	subject='[TEST POST PLS IGNORE]',
	recipients=[Recipient{name='null', address='me@me.com', type=To}]
}
09:55:13.505 ERROR com.foo.bar.Emailer - Error sending email...
com.sun.mail.smtp.SMTPSendFailedException: 530 5.7.1 Client was not authenticated

	at com.sun.mail.smtp.SMTPTransport.issueSendCommand(SMTPTransport.java:2343) ~[javax.mail-1.6.0.jar:1.6.0]
	at com.sun.mail.smtp.SMTPTransport.mailFrom(SMTPTransport.java:1808) ~[javax.mail-1.6.0.jar:1.6.0]
	at com.sun.mail.smtp.SMTPTransport.sendMessage(SMTPTransport.java:1285) ~[javax.mail-1.6.0.jar:1.6.0]
```

To repro this, I use the following `email.properties` file: 
```properties
simplejavamail.transportstrategy=SMTP_TLS
simplejavamail.smtp.host=smtp.foo.bar
simplejavamail.smtp.port=587
simplejavamail.defaults.subject=[TEST POST PLS IGNORE]
simplejavamail.defaults.from.address=service@foo.com
simplejavamail.defaults.to.address=target@foo.com
```

**Comment by bbottema** [comment]
For 6.0.0, I will update the code everywhere to indicate exactly where known exceptions will be thrown. This should clear everything up proper.

**Comment by bbottema** [comment]
I'm working on a one time complete write up of all custom runtime exceptions, but I would like to incorporate an automated check in my work flow, preferably using Maven.

I put out a question on Stackoverflow, but so far it seems it rather uncommon to want what I want.

https://stackoverflow.com/q/56379496/441662

**Comment by bbottema** [comment]
After spending a lot of time thinking about this, I am now reversing course: I had started documenting all known thrown exceptions, but I now think this that this was misguided.

The truth is, even if I document all known runtime exceptions, I will never know about all 3rd party runtime exceptions, nor do I expect the user to be able to handle my own runtime exceptions, otherwise they would not be runtime exceptions, but checked ones.

Moreover, regardless of documented runtime exceptions, the library user should anyway run *any* could -not just Simple Java Mail- in a fashion where an exception might occur. This is to be expected. Either by a try catch on a high level (ie. main running code, batch job runner), or by bubbling it up (ie. web container or REST service HTTP error). This point goes regardless of runtime exceptions being documented or not.

The reason I'm going all principle on this issue now is because as I started documented each and every _own_ thrown runtime exception, I found that almost each and every method is subject to them and in some cases I started having 10+ reasons for a MailException occurring. For example sendEmail() -a method that touched almost all functionality- exploded in known reasons for possible runtime exceptions that may never occur.

So I'm reverting my changes and in fact reducing the documentation to a single line informing about the possibility of a `MailException`.
